### PR TITLE
Factor out sort values from InternalSearchHit

### PR DESCRIPTION
**Pull request by cbuescher**
This adds fromXContent method and unit test for sort values that are part of
InternalSearchHit. In order to centralize serialisation and xContent parsing and
rendering code, move all relevant parts to a new class which can be unit tested
much better in isolation.This is part of the preparation for parsing search
responses on the client side.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/search/internal/SearchSortValues.java
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/search/internal/SearchSortValues.java]: Can you move `writeTo` up here? I like to get reading and writing on the same page if possible.

File: core/src/main/java/org/elasticsearch/search/internal/SearchSortValues.java
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/search/internal/SearchSortValues.java]: I don't think this `if` is needed.

File: core/src/main/java/org/elasticsearch/search/internal/SearchSortValues.java
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/search/internal/SearchSortValues.java]: I think I'd prefer to leave this out and do `assertEquals(this.toString(), other.toString())` in the test.

**Review by nik9000 — APPROVED** [review]
Left a few minor things, LGTM though.

File: core/src/main/java/org/elasticsearch/search/internal/SearchSortValues.java
Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/internal/SearchSortValues.java]: ++

File: core/src/main/java/org/elasticsearch/search/internal/SearchSortValues.java
Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/internal/SearchSortValues.java]: shall these be exceptions? I don't have a strong opinion but I got this same comment from Tanguy in another PR.

File: core/src/main/java/org/elasticsearch/search/internal/SearchSortValues.java
Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/internal/SearchSortValues.java]: shall we check that sortValueFormats is not null? also sortValues?

File: core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java
Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java]: seems like this empty array could be useful, shall we rather have a SearchSortValues.EMPTY default ?

File: core/src/main/java/org/elasticsearch/search/internal/SearchSortValues.java
Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/internal/SearchSortValues.java]: ++

File: core/src/main/java/org/elasticsearch/search/internal/SearchSortValues.java
Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/internal/SearchSortValues.java]: agreed

File: core/src/test/java/org/elasticsearch/search/internal/SearchSortValuesTests.java
Comment (javanna) [inline core/src/test/java/org/elasticsearch/search/internal/SearchSortValuesTests.java]: interesting, more elegant than a switch. I may steal this on :)

**Review by javanna — APPROVED** [review]
left a few comments but I like it

File: core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java]: Can do.

File: core/src/main/java/org/elasticsearch/search/internal/SearchSortValues.java
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/internal/SearchSortValues.java]: Either way is fine. I will change it to what you did in #22082.

**Comment by cbuescher** [comment]
@nik9000 @javanna I added a commit that addresses your last comments. For now I exclude SMILE from the xcontent types being tested because it makes roundtrip testing very hard, even if we only compare the json string representation of the object before and after parsing xContent. The problem is that when SMILE sends float values and we don't explicitely cast them back to floats, they are parsed as `double`. I wrote a small test showing this in isolation here: https://github.com/cbuescher/elasticsearch/blob/smileFloatIssue/core/src/test/java/org/elasticsearch/common/xcontent/SmileFloatTests.java

File: core/src/test/java/org/elasticsearch/search/internal/SearchSortValuesTests.java
Comment (javanna) [inline core/src/test/java/org/elasticsearch/search/internal/SearchSortValuesTests.java]: I did quite some digging around these, I did find that the different parsers for SMILE, CBOR and JSON and YAML have a different way of parsing float, either as float, as double, or as double but with double precision. This makes comparisons a bit cumbersome. The issue you were finding with SMILE is that once you parse a float, it is parsed into a double variable, which has double precision. When you print that out it becomes a double, which reparsed it is not a float anymore. The solution isn't to disable SMILE testing though, I resorted to comparing the map representations, essentially by parsing them back into maps. That also solves eventual ordering issue, as json keys ordering doesn't matter. see https://github.com/elastic/elasticsearch/pull/22082/files#diff-3e73f4760da40a9d55345c94edf833c2R108 . I think that we have to move that assertion to some common place, not sure yet where.

File: core/src/test/java/org/elasticsearch/search/internal/SearchSortValuesTests.java
Comment (javanna) [inline core/src/test/java/org/elasticsearch/search/internal/SearchSortValuesTests.java]: can we use Strings.toString instead? is pretty printing important here?

File: core/src/main/java/org/elasticsearch/search/internal/SearchSortValues.java
Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/internal/SearchSortValues.java]: use the static methods in XContentParserUtils instead

File: core/src/test/java/org/elasticsearch/search/internal/SearchSortValuesTests.java
Comment (cbuescher) [inline core/src/test/java/org/elasticsearch/search/internal/SearchSortValuesTests.java]: I changed this using the common helpers you pointed me to, left TODOs where we need to remove those once they are moved to a common class.

File: core/src/test/java/org/elasticsearch/search/internal/SearchSortValuesTests.java
Comment (cbuescher) [inline core/src/test/java/org/elasticsearch/search/internal/SearchSortValuesTests.java]: done

File: core/src/main/java/org/elasticsearch/search/internal/SearchSortValues.java
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/internal/SearchSortValues.java]: done

**Comment by cbuescher** [comment]
@javanna I hope I adressed your last comments, using two (slightly modified) helper methods from your own PR now but leaving a note to remove these once they are available elsewhere. Not sure if we should hold this PR once those are merged.

File: core/src/main/java/org/elasticsearch/search/internal/SearchSortValues.java
Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/internal/SearchSortValues.java]: you can just do parser.nextToken() as part of this line.

File: core/src/main/java/org/elasticsearch/search/internal/SearchSortValues.java
Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/internal/SearchSortValues.java]: parser.list also supports nested arrays and objects. can they appear as part of sort values?

Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/internal/SearchSortValues.java]: I think we may have the same "problem" that we have in get with stored fields, where we print out potentially anything, and when parsing it back we have to figure out what formats we actually support. Note also that parser list supports binary objects, but it returns byte[] for them, which makes it cumbersome to test them, hence I am assuming that part does not get tested at the moment.

**Review by javanna — CHANGES_REQUESTED** [review]
I left some comments, the main point is about formats supported when parsing stuff back, I think we have to figure out exactly what we support and test it rather than support anything which seems to be what we do at the moment.

File: core/src/main/java/org/elasticsearch/search/internal/SearchSortValues.java
Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/internal/SearchSortValues.java]: that's ok with me

File: core/src/main/java/org/elasticsearch/search/internal/SearchSortValues.java
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/internal/SearchSortValues.java]: This is somewhat limited here due to what is possible to go through writeTo/readFrom. Here this is essential single values, they should all be covered by the tests.

Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/internal/SearchSortValues.java]: I see, but with serialization we can always read whatever we wrote in the same format. With parsing on the other hand, we can write many different types but we can parse only a few e.g. strings, numbers and not much more. That is why the generic methods hide complexity some times, especially because we end up supporting stuff that may not be needed and is not tested. Maybe it's ok in this case though. Can you double check if byte[] can be a potential value here?

**Comment by cbuescher** [comment]
@javanna I updated this using the helper methods introduced in `ElasticsearchAssertions` and `XContentHelper` . What happened to `XContentParserUtils.ensureFieldName()`? I had to go back to my previous version of manual checks.

**Comment by javanna** [comment]
> What happened to XContentParserUtils.ensureFieldName()? I had to go back to my previous version of manual checks.

it was removed as there were no usages for it. Feel free to add it back if it's going to be useful in other places too.

**Review by javanna — APPROVED** [review]
left a small comment, LGTM otherwise

File: core/src/main/java/org/elasticsearch/common/xcontent/XContentParserUtils.java
Comment (javanna) [inline core/src/main/java/org/elasticsearch/common/xcontent/XContentParserUtils.java]: does it make sense to compare <null> to something?

Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/common/xcontent/XContentParserUtils.java]: What do you mean? `current` is either the current tokens field name or "<null>" which I thought serves for printing the error message. I might be missing something.

Comment (javanna) [inline core/src/main/java/org/elasticsearch/common/xcontent/XContentParserUtils.java]: I think if the currentName is null we have a bigger problem. why do we do the comparison even? have you tried passing in fieldName set to <null> ? if the current token is a field name, currentName should not be null ever. I would treat that corner case differently.

**Comment by cbuescher** [comment]
On 5.x with 9404411ad4f623ff89833a5e82a9f352a885bab5

**Pull request merged**
